File: php/src/data-structures/DynamicArray.php
<?php

namespace DataStructures;

use SplFixedArray;

class DynamicArray
{

    /** @var int $size */
    private $size;

    /** @var int $initialCapacity */
    private $initialCapacity;

    /** @var mixed SplFixedArray */
    private $data;

    /**
     * DynamicClass constructor.
     * @param int $initialCapacity
     */
    public function __construct(int $initialCapacity)
    {
        $this->initialCapacity = $initialCapacity;
        //$this->data = []; //Dynamic by nature
        $this->data = new SplFixedArray($initialCapacity);
        $this->size = 0;
    }

    /**
     * @param int $index
     * @param mixed $value
     */
    public function set(int $index, $value): void
    {
        $this->data[$index] = $value;
        $this->size++;
    }

    /**
     * @param int $index
     * @return mixed
     */
    public function get(int $index)
    {
        return $this->data[$index];
    }

    /**
     * @param int $index
     * @param mixed $value
     */
    public function insert(int $index, $value): void
    {
        //Check
        if ($this->size == $this->initialCapacity) {
            $this->resize();
        }

        //Copy UP | Shift
        for ($i = $this->size; $i > $index; $i--) {
            $this->data[$i] = $this->data[$i - 1];
        }

        $this->data[$index] = $value;
        $this->size++;
    }

    /**
     * @param mixed $value
     */
    public function add($value): void
    {
        //Check
        if ($this->size == $this->initialCapacity) {
            $this->resize();
        }

        $this->data[$this->size] = $value;
        $this->size++;
    }

    /**
     * @param int $index
     */
    public function delete(int $index): void
    {
        //copy down [1 2 3] => [1 3 3]  <- shift down
        for ($i = $index; $i < $this->size - 1; $i++) {
            $this->data[$i] = $this->data[$i + 1];
        }

        //clear last
        if ($index == $this->size) {
            $this->data[$index] = null;
        } else {
            $this->data[$this->size - 1] = null;
        }

        $this->size--;
    }

    /**
     * @return int
     */
    public function size(): int
    {
        return $this->size;
    }

    /**
     * @return bool
     */
    public function isEmpty(): bool
    {
        return $this->size() == 0;
    }

    /**
     * @param mixed $value
     * @return bool
     */
    public function contains($value): bool
    {
        for ($i = 0; $i < $this->size; $i++) {
            if ($this->data[$i] === $value) {
                return true;
            }
        }
        return false;
    }

    private function resize(): void
    {
        $newCapacity = $this->initialCapacity * 2;
        $newData = new SplFixedArray($newCapacity);

        for ($i = 0; $i < $this->initialCapacity; $i++) {
            $newData[$i] = $this->data[$i];
        }

        $this->data = $newData;
        $this->initialCapacity = $newCapacity;
    }
}

File: ruby/lib/data-structures/dynamic_array.rb
class DynamicArray

    def initialize(initial_capacity)
        @initial_capacity = initial_capacity
        @size = 0
        #data = [] #Dynamic by nature
        @data = Array.new(initial_capacity) #here too, but... let's try
    end

    def set(index, value)
        @data[index] = value
        increase_size
    end

    def get(index)
        @data[index]
    end

    def size
        @size
    end

    def insert(index, value)
        resize?

        #shift up
        (@size + 1).downto(index).each do |i|
            @data[i] = @data[i - 1]
        end

        @data[index] = value
        increase_size
    end

    def add(value)
        resize?

        @data[@size] = value

        increase_size
    end

    def delete(index)
        #shift down
        (index..(@size - 1)).each do |i|
            @data[i] = @data[i + 1]
        end

        #clear last
        if index == @size
            @data[index] = nil
        else
            @data[@size - 1] = nil
        end

        decrease_size
    end

    def contains(value)
        #@data.include? value
        @data.each do |data|
            if data == value then return true
            end
        end
        false
    end

    private
    def resize?
        if @size == @initial_capacity then resize end
    end

    def resize
        new_capacity = @initial_capacity * 2
        new_data = Array.new new_capacity
        new_data = @data.dup
        @data = new_data
    end

    def increase_size
        @size=@size+1
    end

    def decrease_size
        @size=@size-1
    end


end